llvm/lib/SYCLLowerIR/MutatePrintfAddrspace.cpp

```diff
@@ -0,0 +1,141 @@
+  assert(Literal);
+  Type *LiteralType = Literal->getType();
+  // Generate/find a correct literal
+  auto *CASLiteralType = PointerType::get(LiteralType->getPointerElementType(),
+                                          ConstantAddrspaceID);
+  StringRef CASLiteralName(Literal->getName().str() + "._AS2");
+  return M.getOrInsertGlobal(CASLiteralName, CASLiteralType, [&] {
+    StringRef LiteralValue;
+    getConstantStringInfo(Literal, LiteralValue);
+    GlobalVariable *GV = Builder.CreateGlobalString(
+        LiteralValue, CASLiteralName, ConstantAddrspaceID, &M);
+    GV->setLinkage(GlobalValue::LinkageTypes::InternalLinkage);
+    GV->setUnnamedAddr(GlobalValue::UnnamedAddr::None);
+    return GV;
+  });
+}
+
+static FunctionCallee getCASPrintfFunction(Module &M,
+                                           const Function &GenericASPrintfFunc,
+                                           Type *CASLiteralTy) {
+  auto *CASPrintfFuncTy = FunctionType::get(GenericASPrintfFunc.getReturnType(),
+                                            CASLiteralTy, /*isVarArg=*/true);
+  FunctionCallee CASPrintfFunc =
+      M.getOrInsertFunction("_Z18__spirv_ocl_printfPU3AS2Kcz", CASPrintfFuncTy,
+                            GenericASPrintfFunc.getAttributes());
+  auto *Callee = cast<Function>(CASPrintfFunc.getCallee());
+  Callee->setCallingConv(CallingConv::SPIR_FUNC);
+  Callee->setDSOLocal(true);
+  return CASPrintfFunc;
+}
+
+static void buildCASCall(IRBuilder<> &Builder, FunctionCallee CASPrintfFunc,
+                         Constant *CASLiteral, CallInst *CI) {
+  SmallVector<Value *, 4> CallOperands(CI->data_operands_begin(),
+                                       CI->data_operands_end());
+  CallOperands[0] = CASLiteral;
+  Builder.SetInsertPoint(CI);
+  Builder
+      .CreateCall(CASPrintfFunc, CallOperands, CI->getName().str() + "._AS2")
+      ->setTailCall(true);
+}
+
+PreservedAnalyses
+SYCLMutatePrintfAddrspacePass::run(Module &M, ModuleAnalysisManager &MAM) {
+  size_t ReplacedCallsCount = 0;
+  SmallVector<Function *, 1> FunctionsToDrop;
+
+  for (Function &F : M) {
+    if (!F.getName().contains("__spirv_ocl_printf"))
+      continue;
+    auto *LiteralType = F.getArg(0)->getType();
+    if (LiteralType->getPointerAddressSpace() == ConstantAddrspaceID)
+      // No need to replace the literal type and its printf users
+      continue;
+
+    IRBuilder<> Builder(M.getContext());
+    SmallVector<CallInst *, 8> CallInstsToDrop;
+    for (User *U : F.users()) {
+      if (!isa<CallInst>(U))
+        continue;
+      auto *CI = cast<CallInst>(U);
+      Constant *CASLiteral = getCASLiteral(M, Builder, CI);
+      FunctionCallee CASPrintfFunc =
+          getCASPrintfFunction(M, F, CASLiteral->getType());
+      buildCASCall(Builder, CASPrintfFunc, CASLiteral, CI);
+      CallInstsToDrop.emplace_back(CI);
+      ++ReplacedCallsCount;
+    }
+    for (CallInst *CI : CallInstsToDrop)
+      CI->eraseFromParent();
```

Comment:
`printf` returns a value, which could be used and therefore we can't simply remove the call, we need `replaceAllUsesWith`